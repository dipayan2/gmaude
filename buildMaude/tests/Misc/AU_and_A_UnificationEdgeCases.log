PASS AU_and_A_UnificationEdgeCases (exit status: 0)
