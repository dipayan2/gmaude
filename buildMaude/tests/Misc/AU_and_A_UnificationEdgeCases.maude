***
***	Tests for edge cases of associative-identity and
***	associative unification algorithms.
***

set show timing off .

***(
1/26/21
This example shows that replacing a constrained variable
P with unconstrained variables A B leads to unnecessary
incompleteness in versions before Alpha133.
)

fmod ASSOC-LIST is
  sorts Elt Pair List .
  subsort Elt Pair < List .
  op __ : List List -> List [assoc] .
  op __ : Elt Elt -> Pair [assoc] .
  op rev : List -> List .
  vars A B C X Y Z : List .
  vars P Q R : Pair .
  vars E F G : Elt .
endfm

unify P =? A B /\ X P =? P Y .


***(
1/27/21
This example show that the imperfect splitting of upperbound on
T to give upperbounds of 2, 1, 1 on A, B, C is then tightened
to 1, 1, 1 in the post hoc PigPug::tightenConstraints() pass.
A corner case that shows that constraint tightening is possible
and might be useful in a system of equations.

5/26/21
We now rely on WordLevel::checkAssignmentNormalCase() in the
next WordLevel to do the same tightening on variables that
haven't been eliminated.
)

fmod ASSOC-LIST is
  sorts Elt Pair Triple List .
  subsort Elt < Pair < Triple < List .
  op __ : List List -> List [assoc] .
  op __ : Elt Elt -> Pair [assoc] .
  op __ : Elt Pair -> Triple [assoc] .
  op __ : Pair Elt -> Triple [assoc] .

  op rev : List -> List .
  vars A B C X Y Z : List .
  vars P Q R : Pair .
  vars E F G : Elt .
  vars S T U V : Triple .
endfm

unify T X =? A B C Y .


***(
2/18/21
Demonstrate Alpha133 optimization that avoids identity assignments,
and two cases where a sort-decreasing identity axiom foils the optimization.
)

fmod FOO is
  sort Foo .
  ops 1 a : -> Foo .
  op __ : Foo Foo -> Foo [assoc id: 1] .
vars W X Y Z : Foo .
endfm

unify W =? Y Z .
irred unify W =? Y Z .

fmod FOO2 is
  sorts Lo Hi .
  subsort Lo < Hi .
  op 1 : -> Lo .
  op __ : Hi Hi -> Hi [assoc id: 1] .
vars W X Y Z : Lo .
endfm

unify W =? Y Z .
irred unify W =? Y Z .

fmod FOO3 is
  sorts Lo Hi .
  subsort Lo < Hi .
  op 1 : -> Hi .
  op __ : Hi Hi -> Hi [assoc id: 1] .
  op __ : Lo Lo -> Lo [ditto] .
vars W X : Lo .
vars Y Z : Hi .
endfm

unify W =? Y Z .
irred unify W =? Y Z .


***(
2/18/21
Demonstrate Alpha133 optimization that avoids equality steps in PigPug
and a case where a following equation foils the optimization.
)

fmod FOO is
  sort Foo .
  ops 1 a : -> Foo .
  op __ : Foo Foo -> Foo [assoc id: 1] .
vars W X Y Z : Foo .
endfm

unify W X =? Y Z .
irred unify W =? Y Z .

unify W X =? Y Z /\ W X =? a a .
irred unify W X =? Y Z /\ W X =? a a .
