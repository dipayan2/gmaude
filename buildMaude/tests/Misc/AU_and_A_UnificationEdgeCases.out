==========================================
unify in ASSOC-LIST : P =? A B /\ X P =? P Y .

Unifier 1
P --> #1:Elt #2:Elt
A --> #1:Elt
B --> #2:Elt
X --> #1:Elt #2:Elt #3:List
Y --> #3:List #1:Elt #2:Elt

Unifier 2
P --> #1:Elt #2:Elt
A --> #1:Elt
B --> #2:Elt
X --> #1:Elt #2:Elt
Y --> #1:Elt #2:Elt

Unifier 3
P --> #1:Elt #1:Elt
A --> #1:Elt
B --> #1:Elt
X --> #1:Elt
Y --> #1:Elt
==========================================
unify in ASSOC-LIST : T X =? A B C Y .

Unifier 1
T --> #1:Elt #2:Elt #4:Elt
X --> #5:List #3:List
A --> #1:Elt
B --> #2:Elt
C --> #4:Elt #5:List
Y --> #3:List

Unifier 2
T --> #2:Elt #3:Elt #4:Elt
X --> #1:List
A --> #2:Elt
B --> #3:Elt
C --> #4:Elt
Y --> #1:List

Unifier 3
T --> #1:Elt #4:Pair
X --> #5:List #2:List #3:List
A --> #1:Elt
B --> #4:Pair #5:List
C --> #2:List
Y --> #3:List

Unifier 4
T --> #1:Pair #4:Elt
X --> #5:List #2:List #3:List
A --> #1:Pair
B --> #4:Elt #5:List
C --> #2:List
Y --> #3:List

Unifier 5
T --> #1:Elt #2:Pair
X --> #3:List #4:List
A --> #1:Elt
B --> #2:Pair
C --> #3:List
Y --> #4:List

Unifier 6
T --> #1:Pair #2:Elt
X --> #3:List #4:List
A --> #1:Pair
B --> #2:Elt
C --> #3:List
Y --> #4:List

Unifier 7
T --> #1:Triple
X --> #5:List #2:List #3:List #4:List
A --> #1:Triple #5:List
B --> #2:List
C --> #3:List
Y --> #4:List

Unifier 8
T --> #1:Triple
X --> #2:List #3:List #4:List
A --> #1:Triple
B --> #2:List
C --> #3:List
Y --> #4:List
==========================================
unify in FOO : W =? Y Z .

Unifier 1
W --> #1:Foo #2:Foo
Y --> #1:Foo
Z --> #2:Foo
==========================================
irredundant unify in FOO : W =? Y Z .

Unifier 1
W --> #1:Foo #2:Foo
Y --> #1:Foo
Z --> #2:Foo
==========================================
unify in FOO2 : W =? Y Z .

Unifier 1
W --> #1:Lo
Y --> 1
Z --> #1:Lo

Unifier 2
W --> #1:Lo
Y --> #1:Lo
Z --> 1

Unifier 3
W --> 1
Y --> 1
Z --> 1
==========================================
irredundant unify in FOO2 : W =? Y Z .

Unifier 1
W --> #1:Lo
Y --> 1
Z --> #1:Lo

Unifier 2
W --> #1:Lo
Y --> #1:Lo
Z --> 1
==========================================
unify in FOO3 : W =? Y Z .

Unifier 1
W --> #1:Lo #2:Lo
Y --> #1:Lo
Z --> #2:Lo

Unifier 2
W --> #1:Lo
Y --> 1
Z --> #1:Lo

Unifier 3
W --> #1:Lo
Y --> #1:Lo
Z --> 1
==========================================
irredundant unify in FOO3 : W =? Y Z .

Unifier 1
W --> #1:Lo #2:Lo
Y --> #1:Lo
Z --> #2:Lo

Unifier 2
W --> #1:Lo
Y --> 1
Z --> #1:Lo

Unifier 3
W --> #1:Lo
Y --> #1:Lo
Z --> 1
==========================================
unify in FOO : W X =? Y Z .

Unifier 1
W --> #2:Foo #3:Foo
X --> #1:Foo
Y --> #2:Foo
Z --> #3:Foo #1:Foo

Unifier 2
W --> #1:Foo
X --> #3:Foo #2:Foo
Y --> #1:Foo #3:Foo
Z --> #2:Foo
==========================================
irredundant unify in FOO : W =? Y Z .

Unifier 1
W --> #1:Foo #2:Foo
Y --> #1:Foo
Z --> #2:Foo
==========================================
unify in FOO : W X =? Y Z /\ W X =? a a .

Unifier 1
W --> a
X --> a
Y --> a
Z --> a

Unifier 2
W --> 1
X --> a a
Y --> a
Z --> a

Unifier 3
W --> a a
X --> 1
Y --> a
Z --> a

Unifier 4
W --> a
X --> a
Y --> 1
Z --> a a

Unifier 5
W --> 1
X --> a a
Y --> 1
Z --> a a

Unifier 6
W --> a a
X --> 1
Y --> 1
Z --> a a

Unifier 7
W --> a
X --> a
Y --> a a
Z --> 1

Unifier 8
W --> 1
X --> a a
Y --> a a
Z --> 1

Unifier 9
W --> a a
X --> 1
Y --> a a
Z --> 1
==========================================
irredundant unify in FOO : W X =? Y Z /\ W X =? a a .

Unifier 1
W --> a
X --> a
Y --> a
Z --> a

Unifier 2
W --> 1
X --> a a
Y --> a
Z --> a

Unifier 3
W --> a a
X --> 1
Y --> a
Z --> a

Unifier 4
W --> a
X --> a
Y --> 1
Z --> a a

Unifier 5
W --> 1
X --> a a
Y --> 1
Z --> a a

Unifier 6
W --> a a
X --> 1
Y --> 1
Z --> a a

Unifier 7
W --> a
X --> a
Y --> a a
Z --> 1

Unifier 8
W --> 1
X --> a a
Y --> a a
Z --> 1

Unifier 9
W --> a a
X --> 1
Y --> a a
Z --> 1
Bye.
